### ISSUE TITLE
Strange BEDPP bug

### ISSUE THREAD

**Issue by pbreheny** [pbreheny/grpreg#18]
@YaohuiZeng: I'm seeing a strange bug in the BEDPP rules in this example:

```r
DF <-read.csv(text=RCurl::getURL("https://raw.githubusercontent.com/jonhersh/datasets/master/myDF2.csv"), header=T)

X <- DF[,1:74]
Y <- DF[,75]
groups <- c(1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,2,2,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,5,5,5,5,5,5,5,5,5,6,6,6,7,7,7,7,7,8,8,8,8,8,8,8,8,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9)
groups.factor <- factor(groups, labels = c("head","hh demographics","dwelling","public services","durables","education","region/sector","head LF","hh LF"))
fit <- grpreg(X, Y, group=groups.factor)
```

I don't understand why...the data set looks ordinary yet we're getting `-nan`s, BEDPP is rejecting everything, and we get all 0s for all values of &lambda;.  I have turned BEDPP screening off (2aeacd401e39badfbe7438e13d1a1e5f0086cc33) as of grpreg_3.1-4, but if you can fix it, I would like to turn it back on for 3.1-5.

**Comment by YaohuiZeng** [comment]
Hmm, very strange. The [tests](https://github.com/YaohuiZeng/grpreg/blob/9128644217da05f0e1c1a4a110b1e81bdd3ad549/tests/test-bedpp.R) passed using [my version](https://github.com/YaohuiZeng/grpreg) of grpreg.

**Comment by pbreheny** [comment]
OK, but that version is 17 commits behind the current (CRAN) version of grpreg.  Unless the BEDPP code gets updated to be compatible with the rest of grpreg, it isn't going to be on CRAN.

**Comment by YaohuiZeng** [comment]
Okay, so I dived into the code, and found an issue [here](https://github.com/pbreheny/grpreg/blob/master/src/gdfit_gaussian.c#L171). Due to numerical error, it's possible `n * y_norm_sq - pow(n * lam_max, 2) * K_star` being -0.00000 (in this data case), which leads to RHS being nan.

A minor change like below to that line fixes the bug:

```
    double tmp;
    tmp = n * y_norm_sq - pow(n * lam_max, 2) * K_star;
    if (tmp < 0) {
      tmp = 0.0;
    }
    RHS = 2 * n * lam * lam_max * sqrt(K[g]) - (lam_max - lam) * sqrt(tmp);
```

Another issue though is lambda series produced by my version and your version are different but the beta matrix identical:

```
> fit$lambda
  [1] 1.443376e-01 1.315150e-01 1.198316e-01 1.091861e-01 9.948629e-02 9.064820e-02
  [7] 8.259526e-02 7.525773e-02 6.857204e-02 6.248029e-02 5.692971e-02 5.187223e-02
 [13] 4.726405e-02 4.306524e-02 3.923944e-02 3.575352e-02 3.257727e-02 2.968320e-02
 [19] 2.704622e-02 2.464351e-02 2.245425e-02 2.045948e-02 1.864191e-02 1.698582e-02
 [25] 1.547684e-02 1.410192e-02 1.284915e-02 1.170767e-02 1.066759e-02 9.719909e-03
 [31] 8.856419e-03 8.069639e-03 7.352755e-03 6.699556e-03 6.104386e-03 5.562089e-03
 [37] 5.067969e-03 4.617744e-03 4.207517e-03 3.833732e-03 3.493154e-03 3.182832e-03
 [43] 2.900078e-03 2.642443e-03 2.407696e-03 2.193803e-03 1.998911e-03 1.821334e-03
 [49] 1.659531e-03 1.512103e-03 1.377772e-03 1.255375e-03 1.143851e-03 1.042234e-03
 [55] 9.496448e-04 8.652810e-04 7.884118e-04 7.183715e-04 6.545533e-04 5.964046e-04
 [61] 5.434217e-04 4.951456e-04 4.511582e-04 4.110786e-04 3.745595e-04 3.412846e-04
 [67] 3.109659e-04 2.833405e-04 2.581693e-04 2.352343e-04 2.143367e-04 1.952956e-04
 [73] 1.779461e-04 1.621379e-04 1.477340e-04 1.346097e-04 1.226513e-04 1.117553e-04
 [79] 1.018273e-04 9.278124e-05 8.453881e-05 7.702862e-05 7.018561e-05 6.395051e-05
 [85] 5.826933e-05 5.309284e-05 4.837622e-05 4.407861e-05 4.016279e-05 3.659483e-05
 [91] 3.334385e-05 3.038167e-05 2.768265e-05 2.522340e-05 2.298262e-05 2.094091e-05
 [97] 1.908058e-05 1.738551e-05 1.584103e-05 1.443376e-05
> fit.bedpp$lambda
  [1] 1.118034e-01 1.018711e-01 9.282115e-02 8.457517e-02 7.706175e-02 7.021579e-02
  [7] 6.397802e-02 5.829438e-02 5.311567e-02 4.839702e-02 4.409756e-02 4.018006e-02
 [13] 3.661057e-02 3.335819e-02 3.039474e-02 2.769456e-02 2.523425e-02 2.299251e-02
 [19] 2.094992e-02 1.908878e-02 1.739299e-02 1.584784e-02 1.443996e-02 1.315716e-02
 [25] 1.198831e-02 1.092330e-02 9.952907e-03 9.068718e-03 8.263078e-03 7.529009e-03
 [31] 6.860153e-03 6.250716e-03 5.695419e-03 5.189454e-03 4.728437e-03 4.308376e-03
 [37] 3.925632e-03 3.576889e-03 3.259128e-03 2.969596e-03 2.705786e-03 2.465411e-03
 [43] 2.246391e-03 2.046828e-03 1.864993e-03 1.699312e-03 1.548350e-03 1.410799e-03
 [49] 1.285467e-03 1.171270e-03 1.067218e-03 9.724089e-04 8.860228e-04 8.073110e-04
 [55] 7.355917e-04 6.702438e-04 6.107012e-04 5.564482e-04 5.070148e-04 4.619730e-04
 [61] 4.209326e-04 3.835381e-04 3.494657e-04 3.184201e-04 2.901325e-04 2.643580e-04
 [67] 2.408731e-04 2.194746e-04 1.999771e-04 1.822117e-04 1.660245e-04 1.512753e-04
 [73] 1.378365e-04 1.255914e-04 1.144342e-04 1.042682e-04 9.500532e-05 8.656531e-05
 [79] 7.887509e-05 7.186804e-05 6.548348e-05 5.966611e-05 5.436554e-05 4.953585e-05
 [85] 4.513523e-05 4.112554e-05 3.747206e-05 3.414314e-05 3.110996e-05 2.834624e-05
 [91] 2.582804e-05 2.353354e-05 2.144289e-05 1.953796e-05 1.780226e-05 1.622076e-05
 [97] 1.477975e-05 1.346676e-05 1.227041e-05 1.118034e-05
> all.equal(fit$beta, fit.bedpp$beta, check.attributes = FALSE)
[1] TRUE

```
`fit` is the result from your version after fixing the bug; `fit.bedpp` is the result from my version.

Still checking, but didn't see you change anything related to SetupLambda.

Would love to update my version to be compatible with latest grpreg. Do I just start a pull request?

**Comment by pbreheny** [comment]
Thanks Yaohui!  Fixed by ba89969.
